File: Leetcode/Algorithm/cpp/Word Break.cc
class Solution {
public:
    bool wordBreak(string s, unordered_set<string> &dict) {
        memset(dp, 0, sizeof(dp));
        dp[0] = 1;
        int len = s.length();
        
        for (int i = 0; i <= len; i++) {
            if (dp[i]) {
                string tmp;
                for (int j = i + 1; j <= len; j++) {
                    tmp += s[j - 1];
                    if (dict.find(tmp) != dict.end()) {
                        dp[j] = 1;
                    }
                }
            }
        }
        return dp[len];
    }
    
    static const int SIZE = 1024;
    int dp[SIZE];
};

File: Leetcode/Algorithm/python/Word Break.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        st = set(wordDict)
        n = len(s)
        dp = [0 for i in xrange(n + 1)]
        dp[0] = 1
        
        for i in xrange(n):
            for j in xrange(i + 1):
                snippet = s[j: i + 1]
                if snippet in st:
                    dp[i + 1] |= dp[j]
        return dp[n] == 1
